### DISCUSSION TITLE
Active Directory doesn't work on core 5?

### DISCUSSION THREAD

**Discussion by juliermefelix** [serenity-is/Serenity#5440] [Help]
I configured in appsettings to use Active Directory but it is not working.
do i need to do anything else?
![image](https://user-images.githubusercontent.com/7516443/107081126-fc60e880-67d0-11eb-9aa0-a72bbbe0a507.png)

**Comment by erehtolleh** [comment]
The newer version doesnt come with all the necessary file.  Take a look at this #4372 .

**Reply by LaFong** [reply]
I created all of setup from the wiki, but there are build errors for the added code. Context errors for Log and Config, and errors on the Setting directives. Except for the namespace name change, code is exactly as on the wiki. I am using VS 2022, if that makes a difference. Everything seems to work, otherwise. The namespace and class name looks right, does not seem to be a using directive mssing.

**Severity	Code	Description	Project	File	Line	Suppression State**
Error	CS0246	The type or namespace name 'SettingScopeAttribute' could not be found (are you missing a using directive or an assembly reference?)	Chair.Web	C:\Users\3230999\source\repos\Chair\Chair\Chair.Web\Modules\Administration\User\Authentication\ActiveDirectoryService.cs	41	Active

Error	CS0246	The type or namespace name 'SettingScope' could not be found (are you missing a using directive or an assembly reference?)	Chair.Web	C:\Users\3230999\source\repos\Chair\Chair\Chair.Web\Modules\Administration\User\Authentication\ActiveDirectoryService.cs	41	Active

Error	CS0246	The type or namespace name 'SettingKeyAttribute' could not be found (are you missing a using directive or an assembly reference?)	Chair.Web	C:\Users\3230999\source\repos\Chair\Chair\Chair.Web\Modules\Administration\User\Authentication\ActiveDirectoryService.cs	41	Active

Error	CS0246	The type or namespace name 'SettingKey' could not be found (are you missing a using directive or an assembly reference?)	Chair.Web	C:\Users\3230999\source\repos\Chair\Chair\Chair.Web\Modules\Administration\User\Authentication\ActiveDirectoryService.cs	41	Active

Error	CS0103	The name 'Config' does not exist in the current context	Chair.Web	C:\Users\3230999\source\repos\Chair\Chair\Chair.Web\Modules\Administration\User\Authentication\ActiveDirectoryService.cs	12	Active

Error	CS0103	The name 'Log' does not exist in the current context	Chair.Web	C:\Users\3230999\source\repos\Chair\Chair\Chair.Web\Modules\Administration\User\Authentication\ActiveDirectoryService.cs	22	Active

**Reply by wezmag** [reply]
This is how I implement Active Directory login with **Serene v5.2.4.0**

1. Add NuGet package - System.DirectoryServices.Protocols v5.0.1 (If you migrate to .NET6, then use v6.0.1)
2. Add `LDAPSettings.cs`

``` csharp
namespace MySereneProject.ActiveDirectory
{
    public class LDAPSettings
    {
        public const string SectionKey = "LDAPSettings";
        public string Host { get; set; }
        public int Port { get; set; }
        public string BaseDN { get; set; }
    }
}
```

3. Add `LDAPService.cs`

``` csharp
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using MySereneProject.Administration;
using System;
using System.DirectoryServices.Protocols;
using System.Net;

namespace MySereneProject.ActiveDirectory
{
    public class LDAPService : Administration.IDirectoryService
    {
        protected LDAPSettings LDAPSettings { get; }
        protected ILogger<LDAPService> Logger { get; }
        public LDAPService(IOptions<LDAPSettings> options, ILogger<LDAPService> logger)
        {
            LDAPSettings = options.Value;
            Logger = logger;
        }


        public DirectoryEntry Validate(string username, string password)
        {

            using var ldapConnection = new LdapConnection(new LdapDirectoryIdentifier(LDAPSettings.Host, LDAPSettings.Port));
            var credential = new NetworkCredential(username, password);
            ldapConnection.Credential = credential;
            try
            {
                ldapConnection.Bind();
                var request = new SearchRequest
                {
                    DistinguishedName = LDAPSettings.BaseDN,
                    Filter = $"(sAMAccountName={username})"
                };
                SearchResponse response = (SearchResponse)ldapConnection.SendRequest(request);
                if (response.Entries.Count == 1)
                {
                    var entry = response.Entries[0];
                    return new DirectoryEntry()
                    {
                        Username = entry.Attributes["samaccountname"][0].ToString().Trim(),
                        Email = entry.Attributes["mail"][0].ToString().Trim(),
                        FirstName = entry.Attributes["givenname"][0].ToString().Trim(),
                        LastName = entry.Attributes["sn"][0].ToString().Trim()
                    };
                }
            }
            catch (LdapException lex)
            {
                Logger.LogError(lex, $"LdapException: {lex.Message}, host:{LDAPSettings.Host}, username: {username}");
            }
            catch (Exception ex)
            {
                Logger.LogError(ex, $"Exception: {ex.Message}, host:{LDAPSettings.Host}, username: {username}");
            }

            return null;
        }
    }
}
```

4. Add below lines in `Startup.cs`

``` csharp
services.Configure<LDAPSettings>(Configuration.GetSection(LDAPSettings.SectionKey));
services.AddSingleton<Administration.IDirectoryService, LDAPService>();
```
 
6. Add LDAP settings in `appsettings.json`
_Please change the settings accordingly._
``` json
"LDAPSettings": {
   "Host": "contoso.com",
   "Port": 389,
   "BaseDN": "ou=users,dc=contoso,dc=com"
},
```

**Reply by LaFong** [reply]
That seems to work great. Thank you. I had to add the one line "using MySereneProject.ActiveDirectory;" to the top of Startup.cs.

**Reply by ali-h2010** [reply]
This looks great. i will try it. 
But, how to configure the default permission for the LDAP users?
from LDAP itself?

**Reply by MimmoDany** [reply]
HI
wezmag,
I'm trying to implement your access solution with Active Directory. I have implemented the classes described above in my Serenity solution. I compile without errors, but I didn't understand how to log in for domain users.
Can you explain how to do it?

Thanks in advance
